collectors/iscsi.py

```diff
@@ -0,0 +1,247 @@
+                    lun = LUN(client, tpg_lun)
+                    client.luns[lun._name] = lun
+                else:
+                    lun = client.luns[lun_name]
+
+                lun.refresh(self.cycle)
+
+                self.iops += lun.iops
+                self.read_bytes_per_sec += lun.read_bytes_per_sec
+                self.write_bytes_per_sec += lun.write_bytes_per_sec
+                self.total_bytes_per_sec = self.read_bytes_per_sec + \
+                                           self.write_bytes_per_sec
+
+    def prune(self):
+        """
+        drop child objects held by the instance, that are no longer in the
+        iSCSI config i.e. don't report on old tut!
+        :return:
+        """
+        for client_name in self.clients:
+            client = self.clients[client_name]
+
+            for lun_name in client.luns:
+                lun = client.luns[lun_name]
+                if lun._cycle != self.cycle:
+                    # drop the lun entry
+                    self.logger.debug("pruning LUN '{}'".format(lun_name))
+
+                    del client.luns[lun_name]
+
+            if client._cycle != self.cycle:
+                # drop the client entry
+                self.logger.debug("pruning client '{}'".format(client_name))
+                del self.clients[client_name]
+
+    def dump(self):
+
+        gw_stats = {}
+        client_stats = {}
+
+        for metric in ISCSIGateway.metrics:
+            gw_stats[metric] = getattr(self, metric)
+
+        for client_name in self.clients:
+            client = self.clients[client_name]
+            client_stats.update(client.dump())
+
+        return {"iscsi": {
+                           "gw_name": {self.gateway_name: 0},
+                           "gw_stats": gw_stats,
+                           "gw_clients": client_stats
+                         }
+               }
+
+
+    def _get_so(self):
+        return [so for so in self._root.storage_objects]
+
+    def _get_node_acls(self):
+        return [node for node in self._root.node_acls]
+
+    def _get_tpg_count(self):
+        return len([tpg for tpg in self._root.tpgs])
+
+    def _get_lun_count(self):
+        return len(self._get_so())
+
+    def _get_session_count(self):
+        return len([session for session in self._root.sessions])
+
+    def _get_gateway_name(self):
+        # Only the 1st gateway is considered/supported
+        gw_iqn = [gw.wwn for gw in self._root.targets][0]
+        return gw_iqn.replace('.', '-')
+
+    def _get_client_count(self):
+        return len(self._get_node_acls())
+
+    def _get_capacity(self):
+        return sum([so.size for so in self._get_so()])
+
+    lun_count = property(_get_lun_count)
+
+    tpg_count = property(_get_tpg_count)
+
+    sessions = property(_get_session_count)
+
+    client_count = property(_get_client_count)
+
+    gateway_name = property(_get_gateway_name)
+
+    capacity = property(_get_capacity)
```

Comment:
Note on defining properties: if you're just providing a getter method, you can simply do:
```
@property
def capacity(self):
    return foo
```
Makes things quite a bit nicer to read.